frontend/public/components/nav/PinnedResource.tsx

```diff
@@ -0,0 +1,146 @@
+import { useDrag, useDrop } from 'react-dnd';
+import { useTranslation } from 'react-i18next';
+import * as classNames from 'classnames';
+import { Button } from '@patternfly/react-core';
+import { MinusCircleIcon, GripVerticalIcon } from '@patternfly/react-icons';
+import { K8sModel, modelFor, referenceForModel } from '../../module/k8s';
+import confirmNavUnpinModal from './confirmNavUnpinModal';
+import { NavLinkComponent, ResourceClusterLink, ResourceNSLink, RootNavLink } from './items';
+
+import './PinnedResource.scss';
+
+type PinnedResourceProps = {
+  resource?: string;
+  navResources?: string[];
+  onChange?: (pinnedResources: string[]) => void;
+  idx?: number;
+  draggable?: boolean;
+};
+
+type DraggableButtonProps = {
+  dragRef?: (node) => void | null;
+};
+
+type RemoveButtonProps = {
+  resource?: string;
+  navResources?: string[];
+  onChange?: (pinnedResources: string[]) => void;
+};
+
+export type DragItem = {
+  idx: number;
+  id: string;
+  type: string;
+};
+
+const DraggableButton: React.FC<DraggableButtonProps> = ({ dragRef }) => {
+  const { t } = useTranslation();
+  return (
+    <Button
+      ref={dragRef}
+      className="oc-pinned-resource__drag-button"
+      variant="link"
+      type="button"
+      aria-label={t('public~Drag to reorder')}
+    >
+      <GripVerticalIcon className="oc-pinned-resource__drag-icon" />
+    </Button>
+  );
+};
+
+const RemoveButton: React.FC<RemoveButtonProps> = ({ resource, navResources, onChange }) => {
+  const { t } = useTranslation();
+  const unPin = (e: React.MouseEvent<HTMLButtonElement>, navItem: string) => {
+    e.preventDefault();
+    e.stopPropagation();
+    confirmNavUnpinModal(navItem, navResources, onChange);
+  };
+  return (
+    <Button
+      className="oc-pinned-resource__unpin-button"
+      variant="link"
+      aria-label={t('public~Unpin')}
+      onClick={(e) => unPin(e, resource)}
+    >
+      <MinusCircleIcon className="oc-pinned-resource__delete-icon" />
+    </Button>
+  );
+};
+
+const swapArrayElements = (arr: string[], idx1: number, idx2: number) => {
+  const arrTemp = [...arr];
+  [arrTemp[idx1], arrTemp[idx2]] = [arrTemp[idx2], arrTemp[idx1]];
+  return arrTemp;
+};
+
+const PinnedResource: React.FC<PinnedResourceProps> = ({
+  resource,
+  onChange,
+  navResources,
+  idx,
+  draggable,
+}) => {
+  const { t } = useTranslation();
+  const model = modelFor(resource);
+  const [, drag, preview] = useDrag({
+    item: { type: 'NavItem', id: `NavItem-${idx}`, idx },
+  });
+
+  const [{ isOver }, drop] = useDrop({
+    accept: 'NavItem',
+    collect: (monitor) => ({
+      isOver: monitor.isOver(),
+    }),
+    drop(item: DragItem) {
+      if (item.idx === idx) {
+        return;
+      }
+      onChange(swapArrayElements(navResources, item.idx, idx));
+      // monitor item updated here to avoid expensive index searches.
+      item.idx = idx;
+    },
+  });
+
+  const { apiVersion, apiGroup, namespaced, crd, plural } = model;
```

Comment:
can the model be falsy value here?

Reply:
no, I am handling it in perspective-nav.tsx , line no: 39